src/benchmark_runner.cc

```diff
@@ -124,6 +124,57 @@ void RunInThread(const BenchmarkInstance* b, size_t iters, int thread_id,
   manager->NotifyThreadComplete();
 }
 
+struct AbstractIterationResults {
+  std::vector<internal::ThreadManager::Result> results;
+
+  bool has_error = false;
+  size_t iters = 0;
+  double real_time_used = 0.0;
+  double seconds = 0.0;
+
+  AbstractIterationResults() = default;
+
+  virtual ~AbstractIterationResults() = default;
+
+  // *Append* the new data.
+  void append(internal::ThreadManager::Result result,
+              const benchmark::internal::BenchmarkInstance& b) {
+    has_error |= result.has_error_;
+    iters += result.iterations;
+
+    // Adjust real/manual time stats since they were reported per thread.
+    result.real_time_used /= b.threads;
+    result.manual_time_used /= b.threads;
+
+    // So for how long were we running?
+    real_time_used += result.real_time_used;
+    // Base decisions off of real time if requested by this benchmark.
+    if (b.use_manual_time) {
+      seconds += result.manual_time_used;
+    } else if (b.use_real_time) {
+      seconds += result.real_time_used;
+    } else
+      seconds += result.cpu_time_used;
+
+    results.emplace_back(result);
+  }
+
+  // This is the business logic. You need to decide what this does.
+  virtual void new_result(internal::ThreadManager::Result result,
+                          const benchmark::internal::BenchmarkInstance& b) = 0;
+};
+
+struct IterationAveragedResults final : public AbstractIterationResults {
+  // Just replace all of the existing data.
+  void new_result(internal::ThreadManager::Result result,
+                  const benchmark::internal::BenchmarkInstance& b) final {
+    IterationAveragedResults empty;
+    empty.append(result, b);
+    *this = std::move(empty);
```

Comment:
i know this is almost certainly safe but... ew.

i think i'd prefer something a bit more explicit. like the struct being a handler with the overridden method, and a field that stores the results with an accessor.

Reply:
I'm having a hard time imagining it..
How will it work in the target, more complex, case of #710:
```c++
struct SeparateIterationResults final : public AbstractIterationResults {
  // *Append* the new data.
  void new_result(internal::ThreadManager::Result result,
                  const benchmark::internal::BenchmarkInstance& b) final {
    append(result, b);
  }
};
```